### PR TITLE
Example of exporting BartModel + Beam Search.

### PR DESCRIPTION
**Pull request by fatcat-z**
All new files for the example of exporting BartModel and Beam Search to ONNX file. This example could work with latest version of Huggingface transformers and required no changes to the model code.

### CODE REVIEW

File: examples/onnx/pytorch/translation/bart_onnx/generation_onnx.py [hunk 1]
```diff
@@ -0,0 +1,675 @@
...
+        self.min_length = min_length
+        self.eos_token_id = eos_token_id
+
+    def forward(self, input_ids, scores) -> torch.Tensor:
+        cur_len = input_ids.shape[-1]
+        if cur_len < self.min_length:
+            scores[:, self.eos_token_id] = -float("inf")
+        return scores
+
+class BARTGenerator(torch.nn.Module, GenerationMixin):
```
Comment (patrickvonplaten) [inline examples/onnx/pytorch/translation/bart_onnx/generation_onnx.py:121]: Think I'm fine with overwriting the methods of `generate()` for the sake of this example

File: examples/onnx/pytorch/translation/bart_onnx/generation_onnx.py [hunk 2]
```diff
@@ -0,0 +1,675 @@
...
+        )
+
+        # from generation_utils.py
+        batch_size = input_ids.shape[0]
+
+        length_penalty = self.config.length_penalty
+        num_return_sequences = self.config.num_return_sequences
+        early_stopping = True
+
+        # TODO: initialize BeamSearchScorer
```
Comment (BowenBao) [inline examples/onnx/pytorch/translation/bart_onnx/generation_onnx.py:652]: remove TODO

File: examples/onnx/pytorch/translation/bart_onnx/generation_onnx.py [hunk 3]
```diff
@@ -0,0 +1,675 @@
...
+        next_tokens = torch.zeros((batch_size, num_beams), dtype=torch.long, device=input_ids.device)
+        next_indices = torch.zeros((batch_size, num_beams), dtype=torch.long, device=input_ids.device)
+
+        past: List[torch.Tensor] = []
+        while cur_len < max_length:
+            logits, past = self._decoder_forward(input_ids, encoder_output, attention_mask, past)
+            next_token_logits = logits[:, -1, :]
+
+            # adjust tokens for Bart, *e.g.*
+            # TODO: make script compatible
```
Comment (BowenBao) [inline examples/onnx/pytorch/translation/bart_onnx/generation_onnx.py:581]: remove TODO

File: examples/onnx/pytorch/translation/bart_onnx/generation_onnx.py [hunk 4]
```diff
@@ -0,0 +1,675 @@
...
+        self.logits_processor = MinLengthLogitsProcessorTS(self.config.min_length, self.config.eos_token_id)
+        self.final_logits_weight = model.model.shared.weight
+        self.final_logits_bias = model.final_logits_bias
+        self.decoder_layers = model.config.decoder_layers
+
+    def _trace_modules(self, model):
+        # device = torch.device('cuda') if torch.cuda.is_available() else torch.device('cpu')
+        device = torch.device('cpu')
+
+        # TODO: Try to update the content to a common one.
```
Comment (BowenBao) [inline examples/onnx/pytorch/translation/bart_onnx/generation_onnx.py:143]: resolve TODO

File: examples/onnx/pytorch/translation/run_onnx_exporter.py [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+        "--device",
+        type=str,
+        default='cpu',
+        help="Device where the model will be run",
+    )
+    parser.add_argument("--output_file_path", type=str, default=None, help="Where to store the final ONNX file.")
+
+    args = parser.parse_args()
+
+    # if args.validation_file is not None:
```
Comment (BowenBao) [inline examples/onnx/pytorch/translation/run_onnx_exporter.py:102]: remove unused comments

File: examples/onnx/pytorch/translation/bart_onnx/generation_onnx.py [hunk 6]
```diff
@@ -0,0 +1,675 @@
...
+            True, True, True, True, True, True, True, True, True, True, True, True,
+            True, True, True, True, True, True, True, True, True, True, True, True,
+            True, True, True, True, True, True, True, True, True, True, True, True,
+            True]], device=device, dtype=torch.bool)
+        self.encoder = create_traced_encoder(model.get_encoder(), input_ids, attention_mask)
+        encoder_outputs = model.get_encoder()(input_ids, attention_mask=attention_mask, return_dict=True)
+        decoder = model.model.decoder
+        decoder_outputs = decoder(input_ids, attention_mask, encoder_outputs['last_hidden_state'], None, None, None)
+        self.decoder_no_past = create_traced_decoder(model.model.decoder, input_ids, encoder_outputs['last_hidden_state'], attention_mask)
+        self.decoder_with_past = create_traced_decoder(model.model.decoder, input_ids, encoder_outputs['last_hidden_state'], attention_mask, decoder_outputs[1])
```
Comment (BowenBao) [inline examples/onnx/pytorch/translation/bart_onnx/generation_onnx.py:162]: Use postprocess script to resolve duplicated weight tensors in onnx model.

Comment (fatcat-z) [reply examples/onnx/pytorch/translation/bart_onnx/generation_onnx.py:152]: Done

**Pull request closed without merging**